drivers/lora/sx1276.c

```diff
@@ -0,0 +1,492 @@
+
+int sx1276_write(u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+}
+
+void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_write(addr, buffer[i]);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+	}
+}
+
+void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+	u8_t regval;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_read(addr, &regval, 1);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+
+		buffer[i] = regval;
+	}
+}
+
+void SX1276SetRfTxPower(int8_t power)
+{
+	int ret;
+	u8_t pa_config = 0;
+	u8_t pa_dac = 0;
+
+	ret = sx1276_read(SX1276_REG_PA_CONFIG, &pa_config, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA config");
+		return;
+	}
+
+	ret = sx1276_read(SX1276_REG_PA_DAC, &pa_dac, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA dac");
+		return;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_MAX_POWER_MASK) | 0x70;
+	pa_config &= RF_PACONFIG_PASELECT_MASK;
+
+#if defined CONFIG_PA_BOOST_PIN
+	pa_config |= RF_PACONFIG_PASELECT_PABOOST;
+
+	if (power > 17) {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_ON;
+	} else {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_OFF;
+	}
+
+	if ((pa_dac & RF_PADAC_20DBM_ON) == RF_PADAC_20DBM_ON) {
+		if (power < 5) {
+			power = 5;
+		} else if (power > 20) {
+			power = 20;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 5) & 0x0F);
+	} else {
+		if (power < 2) {
+			power = 2;
+		} else if (power > 17) {
+			power = 17;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 2) & 0x0F);
+	}
+#elif CONFIG_PA_RFO_PIN
+	if (power < -1) {
+		power = -1;
+	} else if (power > 14) {
+		power = 14;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power + 1) & 0x0F);
+#endif
+	ret = sx1276_write(SX1276_REG_PA_CONFIG, pa_config);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA config");
+		return;
+	}
+
+	ret = sx1276_write(SX1276_REG_PA_DAC, pa_dac);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA dac");
+		return;
+	}
+}
+
+static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
+{
+	Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
+
+	Radio.Send(data, data_len);
+
+	return 0;
+}
+
+static void sx1276_rx_done(u8_t *payload, u16_t size, int16_t rssi, int8_t snr)
+{
+	dev_data.rx_buf = payload;
+	dev_data.rx_len = size;
+
+	k_sem_give(&dev_data.data_sem);
+}
+
+static int sx1276_lora_recv(struct device *dev, u8_t *data, u32_t timeout)
+{
+	int ret;
+
+	Radio.SetMaxPayloadLength(MODEM_LORA, 255);
+	Radio.Rx(0);
+
+	if (!timeout) {
+		/* Wait until the data arrives */
+		k_sem_take(&dev_data.data_sem, K_FOREVER);
+	} else {
+		ret = k_sem_take(&dev_data.data_sem, K_MSEC(timeout));
+		if (ret < 0) {
+			LOG_ERR("Receive timeout!");
+			return ret;
+		}
+	}
```

Comment:
I was thinking more about something like this:
```C
ret = k_sem_take(&dev_data.data_sem, timeout == K_FOREVER ? K_FOREVER :
                              timeout == K_NO_WAIT ? K_NO_WAIT : 
                              K_MSEC(timeout));
if (ret < 0) {
    LOG_ERR("Receive timeout!");
    return ret;
}
```
Otherwise, you would introduce a new value for K_FOREVER, which is -1.

Reply:
@alexanderwachter hmm, makes sense!

Reply:
@alexanderwachter Done!

Reply:
Thanks!